common/formatting/layout_optimizer.h

```diff
@@ -0,0 +1,445 @@
+  }
+
+  UnwrappedLine AsUnwrappedLine() const {
+    assert(type_ == LayoutType::kLayoutLine);
+    UnwrappedLine uwline(0, tokens_.begin());
+    uwline.SpanUpToToken(tokens_.end());
+    uwline.SetPartitionPolicy(PartitionPolicyEnum::kAlwaysExpand);
+    return uwline;
+  }
+
+  int Length() const {
+    assert(type_ == LayoutType::kLayoutLine);
+    if (tokens_.size() == 0) {
+      return 0;
+    }
+    int len = 0;
+    for (const auto& token : tokens_) {
+      len += token.before.spaces_required;
+      len += token.Length();
+    }
+    len -= tokens_.front().before.spaces_required;
+    return len;
+  }
+
+  static int UnwrappedLineLength(const UnwrappedLine& uwline) {
+    int len = 0;
+    for (const auto& token : uwline.TokensRange()) {
+      len += token.before.spaces_required;
+      len += token.Length();
+    }
+    len -= uwline.TokensRange().front().before.spaces_required;
+    return len;
+  }
+
+  SpacingOptions SpacingOptionsLayout() const {
+    assert(type_ == LayoutType::kLayoutLine);
+    assert(tokens_.size() > 0);
+    return tokens_.front().before.break_decision;
+  }
+
+  bool MustWrapLayout() const {
+    return SpacingOptionsLayout() == SpacingOptions::MustWrap;
+  }
+
+  bool MustAppendLayout() const {
+    return SpacingOptionsLayout() == SpacingOptions::MustAppend;
+  }
+
+ private:
+  const LayoutType type_;
+
+  const int indentation_;
+
+  /* const */ FormatTokenRange tokens_;
+
+  /* const */ int spaces_before_;
+};
+
+std::ostream& operator<<(std::ostream& stream, const Layout& layout);
+
+using LayoutTree = VectorTree<const Layout>;
+
+class Knot {
+  friend std::ostream& operator<<(std::ostream&, const Knot&);
+
+ public:
+  explicit Knot(int column, int span, float intercept, int gradient,
+                LayoutTree layout, int before_spaces,
+                SpacingOptions break_decision)
+      : column_(column),
+        span_(span),
+        intercept_(intercept),
+        gradient_(gradient),
+        layout_(layout),
+        before_spaces_(before_spaces),
+        break_decision_(break_decision) {}
+
+  Knot(const Knot&) = default;
+  Knot(Knot&&) = default;
+
+  ~Knot() = default;
+
+  // Deleting standard interfaces:
+  Knot() = delete;
+  Knot& operator=(const Knot&) = delete;
+  Knot& operator=(Knot&&) = delete;
+
+  int GetColumn() const { return column_; }
+  int GetSpan() const { return span_; }
+  float GetIntercept() const { return intercept_; }
+  int GetGradient() const { return gradient_; }
+  LayoutTree GetLayout() const { return layout_; }
+  int GetSpacesBefore() const { return before_spaces_; }
+  SpacingOptions GetSpacingOptions() const { return break_decision_; }
+
+  bool MustWrap() const { return break_decision_ == SpacingOptions::MustWrap; }
+
+  bool MustAppend() const {
+    return break_decision_ == SpacingOptions::MustAppend;
+  }
+
+  // Total cost if this knot if placed at column 'm'
+  float ValueAt(int m) const {
+    assert(m >= 0 && m >= column_);
```

Comment:
`CHECK(m >= 0 && m >= column_);`  ... there are more asserts in this file

Reply:
Done